### PR TITLE
ShortcutExpander has been extended in a way that it will examine the …

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**:

ShortcutExpander has been extended in a way that it will examine the server resources in the first place when searching for an alternative name for the resource.

Note that the list returned from the server is ordered and the first match will yield the extended resource's name.

APIResource struct has been extended by the shorName field - at the moment will always be empty no server side logic fill this field.

### CODE REVIEW

**Comment by deads2k** [comment]
Let's break this into two pulls.  One for `ShortcutExpander` which uses an instance variable to hold the anticipated result of the discovery API information and a separate one for the discovery changes.  The `ShorcutExpander` changes require less scrutiny and backwards compatibility consideration.

File: pkg/apis/meta/v1/types.go [hunk 1]
```diff
@@ -433,6 +433,8 @@ type APIResource struct {
 	// verbs is a list of supported kube verbs (this includes get, list, watch, create,
 	// update, patch, delete, deletecollection, and proxy)
 	Verbs Verbs `json:"verbs" protobuf:"bytes,4,opt,name=verbs"`
+	// shortName is the suggested short name of the resource.
+	ShortName string `json:"shortName" protobuf:"bytes,5,opt,name=shortName"`
```
Comment (deads2k) [inline pkg/apis/meta/v1/types.go]: @kubernetes/sig-api-machinery Think this should be a list so people can transition their shortcuts over time?

File: pkg/kubectl/cmd/util/cached_discovery_test.go [hunk 2]
```diff
@@ -85,22 +92,29 @@ func TestNewCachedDiscoveryClient_TTL(t *testing.T) {
...
+func NewFakeDiscoveryClient() *fakeDiscoveryClient {
+	return &fakeDiscoveryClient{ServerResourcesForGroupVersionHandler: nil, ServerResourcesHandler: nil}
 }
 
 type fakeDiscoveryClient struct {
-	groupCalls    int
-	resourceCalls int
-	versionCalls  int
-	swaggerCalls  int
+	GroupCalls    int
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/cached_discovery_test.go]: Why the change?

**Comment by smarterclayton** [comment]
Probably.

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 3]
```diff
@@ -40,13 +42,20 @@ func NewShortcutExpander(delegate meta.RESTMapper, client discovery.DiscoveryInt
 	return ShortcutExpander{All: userResources, RESTMapper: delegate, discoveryClient: client}
 }
 
-func (e ShortcutExpander) getAll() []schema.GroupResource {
+func (e ShortcutExpander) getServerRes(groupVersion string) ([]*metav1.APIResourceList, error) {
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: I'd like to more narrowly scope this return value.  I think we need an order list of (shortname,group) tuples to (resource,group) tuples.  The ordering matters for priority.

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 4]
```diff
@@ -124,17 +133,33 @@ func (e ShortcutExpander) AliasesForResource(resource string) ([]string, bool) {
...
+				if apiRes.ShortName == gvk.Resource {
+					gvk.Resource = apiRes.Name
+					return gvk
+				}
+			}
+		}
+	}
+
+	// fall back to the hardcoded list
+	if expanded, ok := kubectl.ShortForms[gvk.Resource]; ok {
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: I think the shortforms need to be updated to have the same logical ordering and tupled with group.

**Comment by deads2k** [comment]
> k8s-merge-robot assigned jbeda and unassigned deads2k 2 hours ago

@eparis whoa, how do I slap this bot?

**Comment by eparis** [comment]
Don't race the bot...

File: pkg/kubectl/cmd/util/shortcut_restmapper.go (hunk 3 above)
Comment (p0lyn0mial) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: Why we need group in the second tuple (resource, group) ?

File: pkg/kubectl/cmd/util/cached_discovery_test.go (hunk 2 above)
Comment (p0lyn0mial) [inline pkg/kubectl/cmd/util/cached_discovery_test.go]: good catch - no need to export - my original idea was to move the code outside the package.

**Comment by p0lyn0mial** [comment]
regarding the instance variable that will live within `ShortcutExpander` should it be pre-populated ? If yes what data should I put in ?

**Comment by deads2k** [comment]
> regarding the instance variable that will live within ShortcutExpander should it be pre-populated ? If yes what data should I put in ?

A hard coded set of tuples would work to prove the flow/value.

File: pkg/kubectl/cmd/util/shortcut_restmapper.go (hunk 3 above)
Comment (p0lyn0mial) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: pinned to what ? to `ShortcutExpander` ? - do you suspect that the list obtained from the server can change over the lifetime of the `kubectl` process ?

**Comment by deads2k** [comment]
@kubernetes/sig-api-machinery-misc https://github.com/kubernetes/kubernetes/pull/38835 is close to merging.  We should hammer out the discovery API here.

File: pkg/apis/meta/v1/types.go (hunk 1 above)
Comment (deads2k) [inline pkg/apis/meta/v1/types.go]: @smarterclayton @liggitt @lavalamp @caesarxuchao do we want this to be a list or single string?

**Comment by deads2k** [comment]
I just have the one question about list versus single string.  I think I prefer a list, but I don't feel very strongly.

After we get the API, we'll need to update the API installer to gather the information about shortnames.  @p0lyn0mial just how deep into API machinery do you want to go?  It would be separate pull, but somewhere down this path: https://github.com/kubernetes/kubernetes/blob/master/pkg/genericapiserver/genericapiserver.go#L221, we provide a list of resources.  You'd probably have plumb through information in https://github.com/kubernetes/kubernetes/blob/master/pkg/genericapiserver/genericapiserver.go#L52 somehow.

**Comment by p0lyn0mial** [comment]
@deads2k - the deeper the better - would be nice to get the whole picture.

File: pkg/apis/meta/v1/types.go (hunk 1 above)
Comment (lavalamp) [inline pkg/apis/meta/v1/types.go]: short names are for humans and shouldn't appear in any scripts or anything, so should there really be anything to migrate?

**Comment by p0lyn0mial** [comment]
Closed by
https://github.com/kubernetes/kubernetes/pull/40312#issuecomment-275515283
https://github.com/kubernetes/kubernetes/pull/38835#event-914403188
